vip-filesystem/class.vip-filesystem-stream.php

```diff
@@ -0,0 +1,551 @@
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @param   string      $path
+	 * @param   int         $flags
+	 *
+	 * @return  array|bool  The file statistics or false if failed
+	 */
+	public function url_stat( $path, $flags ) {
+		$path = $this->trim_path( $path );
+
+		// Default stats
+		$stats = array (
+			0         => 0,
+			'dev'     => 0,
+			1         => 0,
+			'ino'     => 0,
+			2         => 16895,
+			'mode'    => 16895,
+			3         => 0,
+			'nlink'   => 0,
+			4         => 0,
+			'uid'     => 0,
+			5         => 0,
+			'gid'     => 0,
+			6         => -1,
+			'rdev'    => -1,
+			7         => 0,
+			'size'    => 0,
+			8         => 0,
+			'atime'   => 0,
+			9         => 0,
+			'mtime'   => 0,
+			10        => 0,
+			'ctime'   => 0,
+			11        => -1,
+			'blksize' => -1,
+			12        => -1,
+			'blocks'  => -1,
+		);
+
+		$extension = pathinfo( $path, PATHINFO_EXTENSION );
+		/**
+		 * If the file is actually just a path to a directory
+		 * then return it as always existing. This is to work
+		 * around wp_upload_dir doing file_exists checks on
+		 * the uploads directory on every page load.
+		 *
+		 * Added by Joe Hoyle
+		 *
+		 * Hanif's note: Copied from humanmade's S3 plugin
+		 */
+		if ( ! $extension ) {
+			return $stats;
+		}
+
+		$info = array();
+		$result = $this->client->is_file( $path, $info );
+		if ( is_wp_error( $result ) || $result instanceof \WP_Error ) {
+			trigger_error( $result->get_error_message(), E_USER_WARNING );
+			return false;
+		}
+		if ( ! $result ) {
+			// File not found
+			return false;
+		}
+
+		// Here we should parse the meta data into the statistics array
+		// and then combine with data from `is_file` API
+		// see: http://php.net/manual/en/function.stat.php
+		$stats[2] = $stats['mode'] = 33206; // read+write permissions
+		$stats[7] = $stats['size'] = (int) $info['size'];
+		$stats[8] = $stats['atime'] = (int) $info['mtime'];
+		$stats[9] = $stats['mtime'] = (int) $info['mtime'];
+		$stats[10] = $stats['ctime'] = (int) $info['mtime'];
+
+		return $stats;
+	}
+
+	/**
+	 * This method is called in response to fseek() to determine the current position.
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @return  bool|int    Returns current position or false on failure
+	 */
+	public function stream_tell() {
+		return $this->file ? ftell( $this->file ) : false;
+	}
+
+	/**
+	 * Called in response to rename() to rename a file or directory.
+	 *
+	 * @since   1.0.0
+	 * @access  public
+	 *
+	 * @param   string  $path_from  Path to file to rename
+	 * @param   string  $path_to    New path to the file
+	 *
+	 * @return  bool    True on successful rename
+	 */
+	public function rename( $path_from, $path_to ) {
+		$path_from = $this->trim_path( $path_from );
+		$path_to = $this->trim_path( $path_to );
+
+		// Get original file first
+		// Note: Subooptimal. Should figure out a way to do this without downloading the file as this could
```

Comment:
I guess probably best to add a new endpoint for this? Although, we may want to see how frequently this is used before investing too much time in that.